File: tui/src/scala/tui/widgets/canvas/map.scala
package tui
package widgets
package canvas

sealed trait MapResolution
object MapResolution {
  case object Low extends MapResolution
  case object High extends MapResolution
}

/// Shape to draw a world map with the given resolution and color
case class WorldMap(
    resolution: MapResolution = MapResolution.Low,
    color: Color = Color.Reset
) extends Shape {
  override def draw(painter: Painter): Unit = {
    val data = resolution match {
      case MapResolution.Low  => world.WORLD_LOW_RESOLUTION
      case MapResolution.High => world.WORLD_HIGH_RESOLUTION
    }
    data.foreach { case Point(x, y) =>
      painter.get_point(x, y) match {
        case Some((x, y)) => painter.paint(x, y, color)
        case None         => ()
      }
    }
  }
}

File: jatatui-widgets/src/java/jatatui/widgets/canvas/Map.java
package jatatui.widgets.canvas;

import jatatui.core.style.Color;

/// A world map.
///
/// A world map can be rendered with different [resolutions](MapResolution) and [colors](Color).
///
/// Mirrors `ratatui_widgets::canvas::Map` (v0.30).
public record Map(MapResolution resolution, Color color) implements Shape {

  /// Returns a [Map] with the default resolution ([MapResolution#Low]) and the default color
  /// ([Color#RESET]).
  public static Map empty() {
    return new Map(MapResolution.defaultResolution(), Color.RESET);
  }

  /// Convenience builder updating the [MapResolution].
  public Map withResolution(MapResolution resolution) {
    return new Map(resolution, color);
  }

  /// Convenience builder updating the [Color].
  public Map withColor(Color color) {
    return new Map(resolution, color);
  }

  @Override
  public void draw(Painter painter) {
    for (Coord coord : resolution.data()) {
      painter
          .getPoint(coord.x(), coord.y())
          .ifPresent(p -> painter.paint(p.x(), p.y(), color));
    }
  }
}
